File: frontend/src/mocks/handlers/training.ts
/**
 * MSW handlers for training API.
 */

import { http, HttpResponse, delay } from 'msw';
import {
  createMockTrainingRun,
  generateMockMetricsHistory,
  mockRunStore,
  mockMetricsStore,
} from '../data/training';
import { TRAINING } from '@/lib/api/endpoints';

export const trainingHandlers = [
  // List runs
  http.get(TRAINING.RUNS, async () => {
    await delay(100);

    const runs = Array.from(mockRunStore.values());

    return HttpResponse.json({
      data: runs,
      meta: { request_id: crypto.randomUUID() },
    });
  }),

  // Create run
  http.post(TRAINING.RUNS, async ({ request }) => {
    await delay(200);

    const body = (await request.json()) as { config?: Record<string, unknown>; total_steps?: number };
    const run = createMockTrainingRun({
      config: body.config as unknown as Record<string, unknown>,
      total_steps: body.total_steps ?? 100000,
    });

    mockRunStore.set(run.id, run);
    mockMetricsStore.set(run.id, []);

    return HttpResponse.json(
      {
        data: run,
        meta: { request_id: crypto.randomUUID() },
      },
      { status: 201 }
    );
  }),

  // Get run
  http.get(`${TRAINING.RUNS}/:id`, async ({ params }) => {
    await delay(50);

    const run = mockRunStore.get(params.id as string);

    if (!run) {
      return HttpResponse.json(
        {
          error: 'not_found',
          message: `Run ${params.id} not found`,
        },
        { status: 404 }
      );
    }

    return HttpResponse.json({
      data: run,
      meta: { request_id: crypto.randomUUID() },
    });
  }),

  // Start run
  http.post(`${TRAINING.RUNS}/:id/start`, async ({ params }) => {
    await delay(100);

    const run = mockRunStore.get(params.id as string);

    if (!run) {
      return HttpResponse.json(
        {
          error: 'not_found',
          message: `Run ${params.id} not found`,
        },
        { status: 404 }
      );
    }

    if (run.status === 'training') {
      return HttpResponse.json(
        {
          error: 'conflict',
          message: 'Run is already training',
        },
        { status: 409 }
      );
    }

    run.status = 'training';
    run.start_time = new Date().toISOString();

    // Generate initial metrics
    const metrics = generateMockMetricsHistory(10000, 1000);
    mockMetricsStore.set(run.id, metrics);
    run.current_step = 10000;
    run.metrics_count = metrics.length;

    return HttpResponse.json({
      data: run,
      meta: { request_id: crypto.randomUUID() },
    });
  }),

  // Stop run
  http.post(`${TRAINING.RUNS}/:id/stop`, async ({ params }) => {
    await delay(100);

    const run = mockRunStore.get(params.id as string);

    if (!run) {
      return HttpResponse.json(
        {
          error: 'not_found',
          message: `Run ${params.id} not found`,
        },
        { status: 404 }
      );
    }

    run.status = 'stopped';
    run.end_time = new Date().toISOString();

    return HttpResponse.json({
      data: run,
      meta: { request_id: crypto.randomUUID() },
    });
  }),

  // Delete run
  http.delete(`${TRAINING.RUNS}/:id`, async ({ params }) => {
    await delay(100);

    const existed = mockRunStore.delete(params.id as string);
    mockMetricsStore.delete(params.id as string);

    if (!existed) {
      return HttpResponse.json(
        {
          error: 'not_found',
          message: `Run ${params.id} not found`,
        },
        { status: 404 }
      );
    }

    return new HttpResponse(null, { status: 204 });
  }),

  // Get metrics
  http.get(`${TRAINING.RUNS}/:id/metrics`, async ({ params, request }) => {
    await delay(50);

    const run = mockRunStore.get(params.id as string);

    if (!run) {
      return HttpResponse.json(
        {
          error: 'not_found',
          message: `Run ${params.id} not found`,
        },
        { status: 404 }
      );
    }

    const url = new URL(request.url);
    const offset = parseInt(url.searchParams.get('offset') ?? '0', 10);
    const limit = parseInt(url.searchParams.get('limit') ?? '100', 10);

    const allMetrics = mockMetricsStore.get(run.id) ?? [];
    const paginatedMetrics = allMetrics.slice(offset, offset + limit);

    const totalCount = allMetrics.length;
    const totalPages = Math.ceil(totalCount / limit);

    return HttpResponse.json({
      items: paginatedMetrics,
      pagination: {
        page: Math.floor(offset / limit),
        page_size: limit,
        total_count: totalCount,
        total_pages: totalPages,
        has_next: offset + limit < totalCount,
        has_previous: offset > 0,
      },
    });
  }),
];

File: src/arcs/api/routers/training.py
"""Training API router."""

import logging

from fastapi import APIRouter, HTTPException, Query, status

from arcs.api.dependencies import TrainingServiceDep, RequestId
from arcs.api.schemas.training import (
    StartTrainingRequest,
    TrainingRunResponse,
    TrainingMetricsResponse,
)
from arcs.api.schemas.common import ApiResponse, ApiMeta, PaginatedResponse, PaginationMeta

logger = logging.getLogger(__name__)

router = APIRouter()


@router.post(
    "/runs",
    response_model=ApiResponse[TrainingRunResponse],
    status_code=status.HTTP_201_CREATED,
)
async def create_run(
    request: StartTrainingRequest,
    service: TrainingServiceDep,
    request_id: RequestId,
) -> ApiResponse[TrainingRunResponse]:
    """Create a new training run."""
    try:
        run = service.create_run(
            config=request.config.model_dump(),
            total_steps=request.total_steps,
        )

        return ApiResponse(
            data=TrainingRunResponse(
                id=run.id,
                config=run.config,
                status=run.status,
                current_step=run.current_step,
                total_steps=run.total_steps,
                metrics_count=len(run.metrics),
                start_time=run.start_time,
                end_time=run.end_time,
                error_message=run.error_message,
            ),
            meta=ApiMeta(request_id=request_id),
        )
    except Exception as e:
        logger.exception(f"Failed to create run: {e}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=str(e),
        )


@router.get(
    "/runs",
    response_model=ApiResponse[list[TrainingRunResponse]],
)
async def list_runs(
    service: TrainingServiceDep,
    request_id: RequestId,
) -> ApiResponse[list[TrainingRunResponse]]:
    """List all training runs."""
    runs = service.list_runs()

    return ApiResponse(
        data=[
            TrainingRunResponse(
                id=r.id,
                config=r.config,
                status=r.status,
                current_step=r.current_step,
                total_steps=r.total_steps,
                metrics_count=len(r.metrics),
                start_time=r.start_time,
                end_time=r.end_time,
                error_message=r.error_message,
            )
            for r in runs
        ],
        meta=ApiMeta(request_id=request_id),
    )


@router.get(
    "/runs/{run_id}",
    response_model=ApiResponse[TrainingRunResponse],
)
async def get_run(
    run_id: str,
    service: TrainingServiceDep,
    request_id: RequestId,
) -> ApiResponse[TrainingRunResponse]:
    """Get a training run by ID."""
    run = service.get_run(run_id)

    if not run:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Run {run_id} not found",
        )

    return ApiResponse(
        data=TrainingRunResponse(
            id=run.id,
            config=run.config,
            status=run.status,
            current_step=run.current_step,
            total_steps=run.total_steps,
            metrics_count=len(run.metrics),
            start_time=run.start_time,
            end_time=run.end_time,
            error_message=run.error_message,
        ),
        meta=ApiMeta(request_id=request_id),
    )


@router.post(
    "/runs/{run_id}/start",
    response_model=ApiResponse[TrainingRunResponse],
)
async def start_run(
    run_id: str,
    service: TrainingServiceDep,
    request_id: RequestId,
) -> ApiResponse[TrainingRunResponse]:
    """Start a training run."""
    try:
        started = service.start_run(run_id)

        if not started:
            raise HTTPException(
                status_code=status.HTTP_409_CONFLICT,
                detail="Run is already training",
            )

        run = service.get_run(run_id)

        return ApiResponse(
            data=TrainingRunResponse(
                id=run.id,
                config=run.config,
                status=run.status,
                current_step=run.current_step,
                total_steps=run.total_steps,
                metrics_count=len(run.metrics),
                start_time=run.start_time,
                end_time=run.end_time,
                error_message=run.error_message,
            ),
            meta=ApiMeta(request_id=request_id),
        )
    except ValueError as e:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=str(e),
        )


@router.post(
    "/runs/{run_id}/stop",
    response_model=ApiResponse[TrainingRunResponse],
)
async def stop_run(
    run_id: str,
    service: TrainingServiceDep,
    request_id: RequestId,
) -> ApiResponse[TrainingRunResponse]:
    """Stop a training run."""
    try:
        service.stop_run(run_id)
        run = service.get_run(run_id)

        return ApiResponse(
            data=TrainingRunResponse(
                id=run.id,
                config=run.config,
                status=run.status,
                current_step=run.current_step,
                total_steps=run.total_steps,
                metrics_count=len(run.metrics),
                start_time=run.start_time,
                end_time=run.end_time,
                error_message=run.error_message,
            ),
            meta=ApiMeta(request_id=request_id),
        )
    except ValueError as e:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=str(e),
        )


@router.delete(
    "/runs/{run_id}",
    status_code=status.HTTP_204_NO_CONTENT,
)
async def delete_run(
    run_id: str,
    service: TrainingServiceDep,
) -> None:
    """Delete a training run."""
    if not service.delete_run(run_id):
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Run {run_id} not found",
        )


@router.get(
    "/runs/{run_id}/metrics",
    response_model=PaginatedResponse[TrainingMetricsResponse],
)
async def get_metrics(
    run_id: str,
    service: TrainingServiceDep,
    offset: int = Query(default=0, ge=0),
    limit: int = Query(default=100, ge=1, le=1000),
) -> PaginatedResponse[TrainingMetricsResponse]:
    """Get metrics for a training run with pagination."""
    try:
        run = service.get_run(run_id)
        if not run:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail=f"Run {run_id} not found",
            )

        metrics = service.get_metrics(run_id, offset=offset, limit=limit)
        total_count = len(run.metrics)
        total_pages = (total_count + limit - 1) // limit

        return PaginatedResponse(
            items=[
                TrainingMetricsResponse(
                    step=m.step,
                    policy_loss=m.policy_loss,
                    value_loss=m.value_loss,
                    kl=m.kl,
                    entropy=m.entropy,
                    explained_variance=m.explained_variance,
                    episode_reward=m.episode_reward,
                    episode_length=m.episode_length,
                    timestamp=m.timestamp,
                )
                for m in metrics
            ],
            pagination=PaginationMeta(
                page=offset // limit,
                page_size=limit,
                total_count=total_count,
                total_pages=total_pages,
                has_next=offset + limit < total_count,
                has_previous=offset > 0,
            ),
        )
    except ValueError as e:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=str(e),
        )
